pkg/server/plugin/upstreamauthority/gcpca/gcpca.go

```diff
@@ -0,0 +1,284 @@
+
+	createRequest := privatecapb.CreateCertificateRequest{
+		Parent: parentPath,
+		Certificate: &privatecapb.Certificate{
+			Lifetime: ptypes.DurationProto(validity),
+			CertificateConfig: &privatecapb.Certificate_Config{
+				Config: &privatecapb.CertificateConfig{
+					PublicKey: &privatecapb.PublicKey{
+						Type: privatecapb.PublicKey_PEM_EC_KEY,
+						Key: pem.EncodeToMemory(
+							&pem.Block{
+								Type:  "PUBLIC KEY",
+								Bytes: csrParsed.RawSubjectPublicKeyInfo,
+							},
+						),
+					},
+					SubjectConfig: &privatecapb.CertificateConfig_SubjectConfig{
+						Subject:    &subject,
+						CommonName: csrParsed.Subject.CommonName,
+					},
+					ReusableConfig: &privatecapb.ReusableConfigWrapper{
+						ConfigValues: &privatecapb.ReusableConfigWrapper_ReusableConfig{
+							ReusableConfig: fmt.Sprintf("projects/%s/locations/%s/reusableConfigs/%s", "privateca-data", p.c.Location, "subordinate-mtls-pathlen-0"),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	pcaClient, err := getClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	cresp, err := pcaClient.CreateCertificate(ctx, &createRequest)
+	if err != nil {
+		return nil, err
+	}
+
+	cert, err := pemToX509(cresp.GetPemCertificate())
+	if err != nil {
+		return nil, err
+	}
+
+	certChain := make([]*x509.Certificate, len(cresp.PemCertificateChain))
+	for i, c := range cresp.PemCertificateChain {
+		certChain[i], err = pemToX509(c)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// The last certificate returned from the chain is the root.
+	upstreamRoot := certChain[len(certChain)-1]
+	bundle := x509util.DedupeCertificates([]*x509.Certificate{upstreamRoot})
+
+	derBundle := x509util.RawCertsFromCertificates(bundle)
+
+	// All else comprises the chain (including the issued certificate)
+	fullChain := []*x509.Certificate{cert}
+	fullChain = append(fullChain, certChain...)
+
+	derChain := x509util.RawCertsFromCertificates(fullChain)
+
+	log.Info("Successfully minted new X509")
+	return &upstreamauthority.MintX509CAResponse{
+		X509CaChain:       derChain,
+		UpstreamX509Roots: derBundle,
+	}, nil
+}
+
+// Mints an X.509 CA and responds with the signed X.509 CA certificate
+// chain and upstream X.509 roots. If supported by the implementation,
+// subsequent responses on the stream contain upstream X.509 root updates,
+// otherwise the RPC is completed after sending the initial response.
+//
+// Implementation note:
+// The stream should be kept open in the face of transient errors
+// encountered while tracking changes to the upstream X.509 roots as SPIRE
+// core will not reopen a closed stream until the next X.509 CA rotation.
+func (p *Plugin) MintX509CA(request *upstreamauthority.MintX509CARequest, stream upstreamauthority.UpstreamAuthority_MintX509CAServer) error {
+	p.log.Info("Minting new X509")
+	ctx := stream.Context()
+	minted, err := p.mintX509CA_(p.log, ctx, request.Csr, request.PreferredTtl)
+
```

Comment:
nitpick: remove this newline so that the error producing code and the error check are grouped together